cartography/intel/gcp/crm.py

```diff
@@ -136,41 +137,67 @@ def load_gcp_projects(neo4j_session, data, gcp_update_tag):
-        ON CREATE SET project.firstseen = timestamp()
-        SET project.projectid = {ProjectId},
-        project.projectnumber = {ProjectNumber},
-        project.displayname = {DisplayName},
-        project.lifecyclestate = {LifecycleState},
-        project.lastupdated = {gcp_update_tag}
-        WITH parent, project
-        MERGE (parent)-[r:RESOURCE]->(project)
-        ON CREATE SET r.firstseen = timestamp()
-        SET r.lastupdated = {gcp_update_tag}
-        """
-        neo4j_session.run(
-            query,
-            ParentId=parentid,
-            ProjectId=project['projectId'],
-            ProjectNumber=project['projectNumber'],
-            DisplayName=project.get('name', None),
-            LifecycleState=project.get('lifecycleState', None),
-            gcp_update_tag=gcp_update_tag,
-        )
+        if project.get('parent'):
+            # Project has parents, so set the parent node label to either `GCPOrganization` or `GCPFolder`.
+            if project['parent']['type'] == 'organization':
+                parent_label = 'GCPOrganization'
+            elif project['parent']['type'] == 'folder':
+                parent_label = 'GCPFolder'
+            else:
+                raise NotImplementedError(
+                    "Ingestion of GCP {}s as parent nodes is currently not supported. "
+                    "Please file an issue at https://github.com/lyft/cartography/issues.".format(
+                        project['parent']['type'],
+                    ),
+                )
+            # Create a parent node ID: either `organizations/{id}` or `folders/{id}`.
+            parentid = f"{project['parent']['type']}s/{project['parent']['id']}"
+
+            # Create parent node, create project node, and connect them together.
+            INGEST_PARENT_TEMPLATE = Template("""
+            MERGE (parent:$parent_label{id:{ParentId}})
+            ON CREATE SET parent.firstseen = timestamp()
+
+            MERGE (project:GCPProject{id:{ProjectId}})
+            ON CREATE SET project.firstseen = timestamp()
+            SET project.projectid = {ProjectId},
+            project.projectnumber = {ProjectNumber},
+            project.displayname = {DisplayName},
+            project.lifecyclestate = {LifecycleState},
+            project.lastupdated = {gcp_update_tag}
+
+            MERGE (parent)-[r:RESOURCE]->(project)
+            ON CREATE SET r.firstseen = timestamp()
+            SET r.lastupdated = {gcp_update_tag}
+            """)
+            neo4j_session.run(
+                INGEST_PARENT_TEMPLATE.safe_substitute(parent_label=parent_label),
+                ParentId=parentid,
+                ProjectId=project['projectId'],
+                ProjectNumber=project['projectNumber'],
+                DisplayName=project.get('name', None),
+                LifecycleState=project.get('lifecycleState', None),
+                gcp_update_tag=gcp_update_tag,
+            )
```

Comment:
We could put this in a separate function `_attach_gpc_project(neo4j_session, project, gcp_update_tag)`.

Either that, or make two templates: `project_template` and `parent_template`. Then `neo4j_session.run()` can either run
```
project_template.substitute(
    sub_query=parent_template.substitute(parent_label=parent_label)
)
```
or `project_template.substitute(sub_query='')`

Reply:
I tried out the first option of separating it out into functions - how's this look?  There's a bit of repeated code, but I'm personally ok with repetition if it's easy to understand.

Reply:
I was thinking more like this edit here. We would ingest each project from the loop within `load_gcp_projects()`, but then the loop would also ingest a project's parent by calling `_attach_gcp_project()`.

```
def load_gcp_projects(neo4j_session, data, gcp_update_tag):
    """
    Ingest the GCP projects to Neo4j
    :param neo4j_session: The Neo4j session
    :param data: List of GCP projects; output from crm.get_gcp_projects()
    :param gcp_update_tag: The timestamp value to set our new Neo4j nodes with
    :return: Nothing
    """
    query = Template("""
    MERGE (project:GCPProject{id:{ProjectId}})
    ON CREATE SET project.firstseen = timestamp()
    SET project.projectid = {ProjectId},
    project.projectnumber = {ProjectNumber},
    project.displayname = {DisplayName},
    project.lifecyclestate = {LifecycleState},
    project.lastupdated = {gcp_update_tag}
    """)
    for project in data:
        neo4j_session.run(
            query,
            ProjectId=project['projectId'],
            ProjectNumber=project['projectNumber'],
            DisplayName=project.get('name', None),
            LifecycleState=project.get('lifecycleState', None),
            gcp_update_tag=gcp_update_tag,
        )
        _attach_gcp_project(neo4j_session, project, gcp_update_tag)


def _attach_gcp_project(neo4j_session, project, gcp_update_tag):
    """
    Attach a Projects to their respective parent, as in the Resource Hierarchy.
    """
    if project['parent']['type'] == 'organization':
        parent_label = 'GCPOrganization'
    elif project['parent']['type'] == 'folder':
        parent_label = 'GCPFolder'
    else:
        raise NotImplementedError(
            "Ingestion of GCP {}s as parent nodes is currently not supported. "
            "Please file an issue at https://github.com/lyft/cartography/issues.".format(
                project['parent']['type'],
            ),
        )
    parentid = f"{project['parent']['type']}s/{project['parent']['id']}"
    template = Template("""
    MATCH (project:GCPProject{id:{ProjectId}})

    MERGE (parent:$parent_label{id:{ParentId}})
    ON CREATE SET parent.firstseen = timestamp()

    MERGE (parent)-[r:RESOURCE]->(project)
    ON CREATE SET r.firstseen = timestamp()
    SET r.lastupdated = {gcp_update_tag}
    """)
    neo4j_session.run(
        template.substitute(parent_label=parent_label),
        ProjectId=project['projectId'],
        ParentId=parentid,
        gcp_update_tag=gcp_update_tag,
    )
```

Reply:
Great suggestion.  Just revised it, how's it now?